flare/ase/calculator.py

```diff
@@ -0,0 +1,166 @@
+        local_energies = np.zeros(nat)
+
+        for n in range(nat):
+            chemenv = AtomicEnvironment(struc_curr, n,
+                                        self.gp_model.cutoffs)
+            local_energies[n] = self.gp_model.predict_local_energy(chemenv)
+
+        return np.sum(local_energies)
+
+    def get_forces(self, atoms):
+        if self.use_mapping:
+            return self.get_forces_mgp(atoms)
+        else:
+            return self.get_forces_gp(atoms)
+
+    def get_forces_gp(self, atoms):
+        nat = len(atoms)
+        struc_curr = Structure(np.array(atoms.cell), 
+                               atoms.get_atomic_numbers(),
+                               atoms.positions)
+
+        if self.par:
+            res = predict_on_structure_par(struc_curr, self.gp_model)
+        else:
+            res = predict_on_structure(struc_curr, self.gp_model)
+
+        forces = np.zeros((nat, 3))
+        stds = np.zeros((nat, 3))
+        for n in range(nat):
+            forces[n] = res[n][0]
+            stds[n] = res[n][1]
+
+        self.results['stds'] = stds
+        atoms.get_uncertainties = self.get_uncertainties
+
+        return forces
+
+    def get_forces_mgp(self, atoms):
+        nat = len(atoms)
+        struc_curr = Structure(np.array(atoms.cell), 
+                               atoms.get_atomic_numbers(),
+                               atoms.positions)
+
+        forces = np.zeros((nat, 3))
+        stds = np.zeros((nat, 3))
+        for n in range(nat):
+            chemenv = AtomicEnvironment(struc_curr, n,
+                                        self.mgp_model.GP.cutoffs)
+            f, v = self.mgp_model.predict(chemenv, mean_only=False)
+            forces[n] = f
+            stds[n] = np.sqrt(np.absolute(v))
+
+        self.results['stds'] = stds
+        atoms.get_uncertainties = self.get_uncertainties
+        return forces
+
+    def get_stress(self, atoms):
+        return np.eye(3)
+
+    def calculation_required(self, atoms, quantities):
+        return True
+
+    def get_uncertainties(self):
+        return self.results['stds']
+
+    def train_gp(self, monitor=True):
+        self.gp_model.train(monitor)
+
+    def build_mgp(self, skip=True):
+        # l_bound not implemented
+
+        if skip and (self.curr_step in self.non_mapping_steps):
+            return 1
+
+        # set svd rank based on the training set, grid number and threshold 1000
+        grid_params = self.mgp_model.grid_params
+        struc_params = self.mgp_model.struc_params
+
+        train_size = len(self.gp_model.training_data)
+        rank_2 = np.min([1000, grid_params['grid_num_2'], train_size*3])
+        rank_3 = np.min([1000, grid_params['grid_num_3'][0]**3, train_size*3])
+        grid_params['svd_rank_2'] = rank_2
+        grid_params['svd_rank_3'] = rank_3
+       
+        self.mgp_model = MappedGaussianProcess(self.gp_model, grid_params, struc_params)
+
+def predict_on_atom(params):
```

Comment:
These functions look pretty similar to the ones in predict.py, would it be possible to use those so that we don't repeat code?